core/arch/arm/include/kernel/thread_arch.h

```diff
@@ -29,13 +29,21 @@
 
 struct mobj;
 
+struct thread_pauth_keys {
```

Comment:
why is it plural? this is just a key

Reply:
I'm just moving the definition so I can use it a bit earlier.
I suppose we could rename it if you think it's important, it should probably go into a separate patch though.

Reply:
On the other hand, this is supposed to abstract the PAUTH key handling and PAUTH has multiple keys.

Reply:
I think naming single key in its plural form might be confusing when reading the code; and when it comes to nitpicks I'd say it is an important one easy to address. For me, personally, functionality comes first but in my experience with OP-TEE these details have always mattered hence my point. 
I suppose it makes sense renaming it in a different PR too.

Reply:
So you mean depending on how many keys we abstract away we need to change the name of the abstraction?

I can add a comment explaining what's the deal with the pauth keys, I hope that should be more helpful.

Reply:
Sure the multiplicity in this case should be give by the name of the variable not the structure definition.
multiple keys would be  `struct thread_pauth_key *keys` or a pointer to the single key would be `struct thread_pauth_key *key`

Otherwise why not have `struct thread_vector_tables,` `struct thread_user_vfp_states`  and so forth.

Reply:
In fact
```
	regs->apiakey_hi = keys->hi;
	regs->apiakey_lo = keys->lo;
```

regs surely points to a structure containing a set of registers, where as keys points to a structure containing a single key

so in fact

```
struct user_mode_ctx {
	struct vm_info vm_info;
	struct vm_paged_region_head *regions;
#if defined(CFG_WITH_VFP)
	struct thread_user_vfp_state vfp;
#endif
#if defined(CFG_TA_PAUTH)
	struct thread_pauth_keys keys;
#endif
	struct ts_ctx *ts_ctx;
	uaddr_t entry_func;
	uaddr_t dump_entry_func;
#ifdef CFG_FTRACE_SUPPORT
	uaddr_t ftrace_entry_func;
#endif
	uaddr_t dl_entry_func;
	uaddr_t ldelf_stack_ptr;
	bool is_32bit;
	bool is_initializing;
	vaddr_t stack_ptr;
};
```

should also have ` struct thread_pauth_key key;` instead

Reply:
Unless you expect to add more keys to the structure sometime in the future of course
```
struct thread_pauth_keys {
	uint64_t akey_hi;
	uint64_t akey_lo;
	uint64_t bkey_hi;
	uint64_t bkey_lo;
       [....]
};
```

Reply:
quickly coming back to this just to wrap it up from my end, I think you should at least change to a way that is more specific, maybe also add a comment


```
/* OP-TEE only uses PAUTH_AKEY */
struct thread_pauth_keys {
	uint64_t akey_hi;
	uint64_t akey_lo;
};
```

Reply:
Fair enough, I'll update.